testsuite/tests/typing-misc/disambiguate_principality.ml

```diff
@@ -274,12 +274,6 @@ let u = function
 ;;
 [%%expect{|
 val u : M.r ref -> int = <fun>
-|}, Principal{|
-Line 3, characters 7-10:
-3 |     !x.lbl
-           ^^^
-Warning 18: this type-based field disambiguation is not principal.
-val u : M.r ref -> int = <fun>
```

Comment:
I'm slightly confused here, why is it that `!x.lbl` does not warn about principality anymore but `x := { lbl = 4 }` does? Basically I don't understand why the test above still fails in principal mode, given that I would expect `(_ : t) as x` and `(x : t)` to be equivalent patterns when `t` contains no flexible inference variable.

Reply:
> I would expect `(_ : t) as x` and `(x : t)` to be equivalent patterns when t contains no flexible inference variable.

And indeed they are, if you look at examples `r` and `s` a few lines above, you'll notice the warning is emitted in the same situations:
```ocaml
let r arg =
  match arg with
  | (x : M.r ref) ->
    !x.lbl
;;
[%%expect{|
val r : M.r ref -> int = <fun>
|}]

let s arg =
  match arg with
  | (x : M.r ref) ->
    x := { lbl = 4 }
;;
[%%expect{|
val s : M.r ref -> unit = <fun>
|}, Principal{|
Line 4, characters 9-20:
4 |     x := { lbl = 4 }
             ^^^^^^^^^^^
Warning 18: this type-based record disambiguation is not principal.
val s : M.r ref -> unit = <fun>
|}]
```

The warning here has nothing to do with the type of the pattern, it's imputable to the way applications are typed. Indeed the following don't warn:
```ocaml
let s' arg =
  match arg with
  | (x : M.r ref) ->
    x.contents <- { lbl = 4 }
;;
[%%expect{|
val s' : M.r ref -> unit = <fun>
|}]

let s'' arg =
  match arg with
  | (x : M.r ref) ->
    ((:=) : _ -> M.r -> unit) x { lbl = 4 }
;;
[%%expect{|
val s'' : M.r ref -> unit = <fun>
|}]
```
but this does:
```ocaml
let s''' (arg : M.r ref) =
  arg := { lbl = 4 }
;;
[%%expect{|
val s''' : M.r ref -> unit = <fun>
|}, Principal{|
Line 2, characters 9-20:
2 |   arg := { lbl = 4 }
             ^^^^^^^^^^^
Warning 18: this type-based record disambiguation is not principal.
val s''' : M.r ref -> unit = <fun>
|}]
```